### PR TITLE
Multi-protocol architecture

### PR DESCRIPTION
**Pull request by Typiqally**
To support multiple communication protocols, an architecture must be set in place to support, which allows for scalability when introducing new protocols. This means that more protocols can be added more easily to the application, which results in an increased number of students that can use the platform.

### CODE REVIEW

File: app/src/main/java/nl/quantifiedstudent/watch/activity/BluetoothPairActivity.kt [hunk 1]
```diff
@@ -49,14 +52,19 @@ class BluetoothPairActivity : AppCompatActivity() {
     private val bluetoothScanResults = mutableListOf<ScanResult>()
 
     private val bluetoothScanResultAdapter: BluetoothScanResultAdapter by lazy {
-        BluetoothScanResultAdapter(bluetoothScanResults) { device ->
-            if (device == null) return@BluetoothScanResultAdapter
+        BluetoothScanResultAdapter(bluetoothScanResults) { device, record ->
```
Comment (NealGeilen) [inline app/src/main/java/nl/quantifiedstudent/watch/activity/BluetoothPairActivity.kt:55]: What is a record? It is an ambiguous definition

File: app/src/main/java/nl/quantifiedstudent/watch/binary/VarInt.kt [hunk 2]
```diff
@@ -0,0 +1,285 @@
+package nl.quantifiedstudent.watch.binary
+
+import nl.quantifiedstudent.watch.extensions.shl
+import java.io.IOException
+import java.io.InputStream
+import java.io.OutputStream
+import java.nio.ByteBuffer
+
+
+object VarInt {
```
Comment (NealGeilen) [inline app/src/main/java/nl/quantifiedstudent/watch/binary/VarInt.kt:10]: Every function is well documented. But the name of the object and the function names are just very ambiguous. The thing a function does needs a description in the name.

**Review by NealGeilen — CHANGES_REQUESTED** [review]

File: app/src/main/java/nl/quantifiedstudent/watch/binary/VarInt.kt [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+     * @param src the ByteBuffer to get the var int from
+     * @return The integer value of the decoded varint
+     */
+    fun getVarInt(src: ByteBuffer): Int {
+        var tmp: Int
+        if (src.get().also { tmp = it.toInt() } >= 0) {
+            return tmp
+        }
+        var result = tmp and 0x7f
+        if (src.get().also { tmp = it.toInt() } >= 0) {
```
Comment (koenjanssenxyz) [inline app/src/main/java/nl/quantifiedstudent/watch/binary/VarInt.kt:79]: Perhaps you could refactor this a bit. Only the number next to shl changes in each if-statement. There might be a more elegant solution.

File: app/src/main/java/nl/quantifiedstudent/watch/binary/VarInt.kt [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+     *
+     * @param src the ByteBuffer to get the var int from
+     * @return The integer value of the decoded long varint
+     */
+    fun getVarLong(src: ByteBuffer): Long {
+        var tmp: Long
+        if (src.get().also { tmp = it.toLong() } >= 0) {
+            return tmp
+        }
+        var result = tmp and 0x7f
```
Comment (koenjanssenxyz) [inline app/src/main/java/nl/quantifiedstudent/watch/binary/VarInt.kt:214]: As mentioned above

**Review by kn4a-com — APPROVED** [review]
Nice job! Looks good. I added some suggestions. Feel free to reach out to me. I'll approve the PR in its current state.

File: app/src/main/java/nl/quantifiedstudent/watch/binary/VarInt.kt (hunk 2 above)
Comment (Typiqally) [reply app/src/main/java/nl/quantifiedstudent/watch/binary/VarInt.kt:10]: This class is taken from another API, which perhaps should be better documented. I will create my own implementation for this class to circumvent the ambiguity.

**Review by kn4a-com — APPROVED** [review]
Much better! Great work.

**Review by NealGeilen — APPROVED** [review]

**Pull request merged**